src/crypto.rs

```diff
@@ -1,158 +1,240 @@
+                    Cipher::Aes256Cbc(key)
+                }
+            },
+        };
+
+        Ok(cipher)
+    }
 }
 
-impl Key {
+impl Cipher {
+    pub fn builder() -> CipherBuilder {
+        CipherBuilder::default()
+    }
+
     /// Returns the length of the key in bits.
-    #[allow(clippy::len_without_is_empty)]
-    pub fn len(&self) -> usize {
+    pub fn bits(&self) -> usize {
         match self {
-            Self::Key128(_) => 128,
-            Self::Key256(_) => 256,
+            Self::Aes128Cbc(_) => 128,
+            Self::Aes256Cbc(_) => 256,
+        }
+    }
+
+    pub fn encoding(&self) -> String {
+        format!("cipher+{}", self.algorithm())
+    }
+
+    pub fn algorithm(&self) -> String {
+        format!("aes-{}-cbc", self.bits())
+    }
+
+    pub(crate) fn block_size(&self) -> usize {
+        // aes blocksize is 128 bits
+        16
+    }
+
+    pub(crate) fn encrypt_with_iv(&self, iv: &[u8], data: &[u8]) -> Result<Vec<u8>> {
+        // generate a random IV if one isn't provided.
+
+        // create a buffer big enough to store the data + padding.
+        let blocks = data.len() / self.block_size() + 1;
+        let mut buf = vec![0u8; blocks * self.block_size()];
+
+        // copy the data into the buffer.
+        buf[..data.len()].copy_from_slice(data);
+
+        // encrypt the data.
+        let encrypted = self.encrypt_raw(iv, &mut buf, data.len())?;
+
+        // return the encrypted data prefixed with the IV.
+        Ok([iv, encrypted].concat())
+    }
+
+    /// Decrypt the data using AES-CBC with PKCS7 padding.
+    pub fn decrypt(&self, data: &mut [u8]) -> Result<Vec<u8>> {
+        if data.len() % self.block_size() != 0 || data.len() < self.block_size() {
+            return Err(error!(
+                40013,
+                format!(
+                    "invalid cipher message data; unexpected length: {}",
+                    data.len()
+                )
+            ));
         }
+        let (iv, buf) = data.split_at_mut(self.block_size());
+        let decrypted = self.decrypt_raw(iv, buf)?;
+        Ok(decrypted.to_vec())
     }
 
     /// Encrypts the given data using AES-CBC with Pkcs7 padding.
-    pub fn encrypt<'a>(&self, iv: &[u8], buf: &'a mut [u8], pos: usize) -> Result<&'a [u8]> {
+    fn encrypt_raw<'a>(&self, iv: &[u8], buf: &'a mut [u8], len: usize) -> Result<&'a [u8]> {
         let iv = GenericArray::from_slice(iv);
         match self {
-            Self::Key128(key) => Cbc::new(Aes128::new(&key.0), iv).encrypt(buf, pos),
-            Self::Key256(key) => Cbc::new(Aes256::new(&key.0), iv).encrypt(buf, pos),
+            Self::Aes128Cbc(key) => {
+                Aes128CbcEnc::new(key.into(), iv).encrypt_padded_mut::<Pkcs7>(buf, len)
+            }
+            Self::Aes256Cbc(key) => {
+                Aes256CbcEnc::new(key.into(), iv).encrypt_padded_mut::<Pkcs7>(buf, len)
+            }
         }
-        .map_err(Into::into)
+        .map_err(|_| unimplemented!())
```

Comment:
Why not return our own error here?

Reply:
This was placeholder stuff when refactoring, thanks.